### ISSUE TITLE
Question: Use of yjs

### ISSUE THREAD

**Issue by JannikStreek** [ether/etherpad-next#68]
## Context

I proposed to @SamTV12345 in a private chat the use or at least a test of [yjs](https://yjs.dev). It's a library which handles concurrent updates on a datasource and could be a good fit for the project. Additionally or even more interestingly it's very easy to integrate in existing editors, see https://yjs.dev/#demos.

By default it uses peer-to-peer connections which isn't very fitting for etherpad. But a plugin exists to leverage websockets. See https://github.com/yjs/y-websocket . A rough server implementation can be seen here: https://github.com/yjs/y-websocke/blob/master/bin/server.js

It would bring research-grade-level conflict resolution for concurrent updates within the editor and, in addition, many existing editor integrations. At least it should be considered to be used as so much comes essentially for free.

EDIT: Also available as ready-to-go backends:
- https://github.com/drifting-in-space/y-sweet
- https://github.com/yjs/y-redis , discussion: https://discuss.yjs.dev/t/y-redis-an-alternative-backend-to-y-websocket/2509/11

## Tasks

I guess I makes sense to first start a small prototype and see how it goes? Especially before integrating the existing server setup.

**Comment by AugustinMauroy** [comment]
In my opinion, the architecture of the pad should be made up of small components:

pad itsel:
```tsx
const Pad:FC = ()=> {
  //logic here
  return (
    <EditableArea/>
  );
}
```

button for bold

pad itsel:
```tsx
const Pad:FC = ()=> {
  const handleBold = () => {}
  return (
    <button onclick={handleBold}//>
  );
}
```

so that each element has its own component and that makes sense

**Comment by SamTV12345** [comment]
Can't we use yjs and still have our own editor? As far as I understood it is quite customizable and we would have a standard implemented than doing something custom.

**Comment by JannikStreek** [comment]
> Can't we use yjs and still have our own editor? As far as I understood it is quite customizable and we would have a standard implemented than doing something custom.

sure, I think thats also possible or at least I can't think of reasons why not. It's just another advantage, that you can integrate existing editors rather quickly.

**Comment by JannikStreek** [comment]
Only the backend data model will be different as the shared data types won't be compatible the existing model - I guess.

**Comment by AugustinMauroy** [comment]
> Only the backend data model will be different as the shared data types won't be compatible the existing model - I guess.

We can add an adaptator on etherpad-next v1.X to allow read from old format. What do you thinks.

**Comment by JannikStreek** [comment]
> > Only the backend data model will be different as the shared data types won't be compatible the existing model - I guess.
> 
> We can add an adaptator on etherpad-next v1.X to allow read from old format. What do you thinks.

All possible, first we would need to test yjs. Then we can decide if we want to use it and how to proceed later with old pads. At least that would be my suggestion. As this will be a complete rewrite, I guess it would also be fine to convert old pads during migration.

**Comment by SamTV12345** [comment]
> > > Only the backend data model will be different as the shared data types won't be compatible the existing model - I guess.
> > 
> > 
> > We can add an adaptator on etherpad-next v1.X to allow read from old format. What do you thinks.
> 
> All possible, first we would need to test yjs. Then we can decide if we want to use it and how to proceed later with old pads. At least that would be my suggestion. As this will be a complete rewrite, I guess it would also be fine to convert old pads during migration.

Absolutely. If you can build a prototype with yjs for evaluation that would be awesome.

**Comment by JannikStreek** [comment]
Ok, I will build a small prototype the next couple of days :)

**Comment by JannikStreek** [comment]
@AugustinMauroy what's you current plan of starting the websocket backend? As I understand it, nextJS doesn't currently support a web socket server backend. But you can add a custom server script, see https://nextjs.org/docs/pages/building-your-application/configuring/custom-server

As there is already a socket folder in app, how did you start it?
